### PR TITLE
feat(coderd): batch agent stats inserts

### LINKED ISSUE #8063: batch  insert workspace agent stat updates
# Problem

Each running workspace agent reports stats every 30 seconds by default via POST to `/api/v2/workspaceagents/me/report-stats`. This handler synchronously persists this statistic to the database.

With large numbers of agents, this can potentially cause large numbers of transactions. With 10,000 workspace agents, this would equate to roughly 333.4 requests per second, each of which would kick off a database transaction.

# Proposed Solution

We can reduce this overhead by batching updates to workspace agent stats.
- Alter InsertWorkspaceAgentStat to persist multiple rows
- Store a batch of workspace agent stats with a limited size (e.g. 10)
- Insert the current batch every batch interval (e.g. 1s) or once the current batch reaches capacity, whichever happens first.
- It may be prudent to add a limited retry on batch insertion in order to avoid dropping stats.

### PR DESCRIPTION
**Pull request by johnstcn**
This PR adds support for batching inserts to the `workspace_agents_stats` table.

Closes https://github.com/coder/coder/issues/8063

### CODE REVIEW

File: cli/server.go [hunk 1]
```diff
@@ -813,6 +814,21 @@ func (r *RootCmd) Server(newAPI func(context.Context, *coderd.Options) (*coderd.
 				options.SwaggerEndpoint = cfg.Swagger.Enable.Value()
 			}
 
+			batchStatsTicker := time.NewTicker(29 * time.Second) // Hard-coding.
```
Comment (johnstcn) [inline cli/server.go:817]: review: Would this be better set at a lower interval, or made configurable?

File: coderd/coderdtest/coderdtest.go [hunk 2]
```diff
@@ -241,6 +244,22 @@ func NewOptions(t testing.TB, options *Options) (func(http.Handler), context.Can
 	if options.FilesRateLimit == 0 {
 		options.FilesRateLimit = -1
 	}
+	if options.StatsBatcher == nil {
+		batchStatsTicker := time.NewTicker(time.Hour)
+		t.Cleanup(batchStatsTicker.Stop)
+		options.StatsBatcher, err = batchstats.New(
+			batchstats.WithStore(options.Database),
+			// Some tests rely on being able to read stats immediately after writing.
+			batchstats.WithBatchSize(1),
```
Comment (johnstcn) [inline coderd/coderdtest/coderdtest.go:253]: reivew: I'm explicitly setting a low batch size for unit tests here.
Alternatively, I can set to the default batch size and pass in a Batcher with a small batch size in the specific unit tests that need it. Thoughts?

File: coderd/database/dbauthz/dbauthz.go [hunk 3]
```diff
@@ -1997,6 +1997,14 @@ func (q *querier) InsertWorkspaceAgentStat(ctx context.Context, arg database.Ins
 	return q.db.InsertWorkspaceAgentStat(ctx, arg)
 }
 
+func (q *querier) InsertWorkspaceAgentStats(ctx context.Context, arg database.InsertWorkspaceAgentStatsParams) error {
+	if err := q.authorizeContext(ctx, rbac.ActionCreate, rbac.ResourceSystem); err != nil {
+		return err
+	}
```
Comment (johnstcn) [inline coderd/database/dbauthz/dbauthz.go:2003]: review: Doing this as a system-level function. This shouldn't ever need to be called by users directly.

File: coderd/database/dbfake/dbfake.go [hunk 4]
```diff
@@ -2701,8 +2701,12 @@ func (q *FakeQuerier) GetWorkspaceAgentStats(_ context.Context, createdAfter tim
...
 
 	statByAgent := map[uuid.UUID]database.GetWorkspaceAgentStatsRow{}
-	for _, agentStat := range latestAgentStats {
-		stat := statByAgent[agentStat.AgentID]
+	for agentID, agentStat := range latestAgentStats {
+		stat := statByAgent[agentID]
+		stat.AgentID = agentStat.AgentID
+		stat.TemplateID = agentStat.TemplateID
+		stat.UserID = agentStat.UserID
+		stat.WorkspaceID = agentStat.WorkspaceID
```
Comment (johnstcn) [inline coderd/database/dbfake/dbfake.go:2709]: review: this was bugged before.

File: coderd/database/queries/workspaceagentstats.sql [hunk 5]
```diff
@@ -22,6 +22,46 @@ INSERT INTO
...
+		connection_median_latency_ms
+	)
+SELECT
+	unnest(@id :: uuid[]) AS id,
+	unnest(@created_at :: timestamptz[]) AS created_at,
+	unnest(@user_id :: uuid[]) AS user_id,
+	unnest(@workspace_id :: uuid[]) AS workspace_id,
+	unnest(@template_id :: uuid[]) AS template_id,
+	unnest(@agent_id :: uuid[]) AS agent_id,
+	jsonb_array_elements(@connections_by_proto :: jsonb) AS connections_by_proto,
```
Comment (johnstcn) [inline coderd/database/queries/workspaceagentstats.sql:53]: review: `connections_by_proto` needs to be handled as `jsonb[]` and not `jsonb[]` as that ends up being passed to `pq.Array` which turns it into something like `{{123,34,106,34,...125}}`.

**Review by johnstcn — COMMENTED** [review]
notes:

File: coderd/coderd.go [hunk 6]
```diff
@@ -287,6 +291,11 @@ func New(options *Options) *API {
 		v := schedule.NewAGPLUserQuietHoursScheduleStore()
 		options.UserQuietHoursScheduleStore.Store(&v)
 	}
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (mtojek) [inline coderd/coderd.go:294]: nit: why did you move it here?

File: coderd/database/queries/workspaceagentstats.sql (hunk 5 above)
Comment (mtojek) [reply coderd/database/queries/workspaceagentstats.sql:53]: :) 

>  jsonb[] and not jsonb[]

File: coderd/database/dbfake/dbfake.go (hunk 4 above)
Comment (mtojek) [reply coderd/database/dbfake/dbfake.go:2709]: Good finding 👍 Will it be covered with tests now?

File: coderd/database/dbauthz/dbauthz.go (hunk 3 above)
Comment (mtojek) [reply coderd/database/dbauthz/dbauthz.go:2003]: Oh, shouldn't we add a message for linter informing why it is system-level access?

File: coderd/coderd.go [hunk 7]
```diff
@@ -287,6 +291,11 @@ func New(options *Options) *API {
...
+
+	if options.StatsBatcher == nil {
+		panic("developer error: options.StatsBatcher is nil")
```
Comment (mtojek) [inline coderd/coderd.go:297]: Is there a reason why `New(options)` can't define a default instance?

File: coderd/coderdtest/coderdtest.go (hunk 2 above)
Comment (mtojek) [reply coderd/coderdtest/coderdtest.go:253]: Which option will make unit tests execute faster?

File: coderd/coderdtest/coderdtest.go [hunk 8]
```diff
@@ -241,6 +244,22 @@ func NewOptions(t testing.TB, options *Options) (func(http.Handler), context.Can
...
+			// Avoid cluttering up test output.
+			batchstats.WithLogger(slog.Make(sloghuman.Sink(io.Discard))),
+			batchstats.WithTicker(batchStatsTicker.C),
+		)
+		require.NoError(t, err, "create stats batcher")
+		ctx, cancel := context.WithCancel(context.Background())
+		t.Cleanup(cancel)
+		go options.StatsBatcher.Run(ctx)
```
Comment (mtojek) [inline coderd/coderdtest/coderdtest.go:261]: Do we need some `closeFunc` here to gracefully shut it down?

File: cli/server.go (hunk 1 above)
Comment (mtojek) [reply cli/server.go:817]: First of all, 29 sec looks magical :) You may need to comment on why is it exactly this.

I didn't look deeper into the review yet, but you may want to introduce 2 limits: timeout and buffer size are getting full.

File: cli/server.go [hunk 9]
```diff
@@ -813,6 +814,26 @@ func (r *RootCmd) Server(newAPI func(context.Context, *coderd.Options) (*coderd.
...
+			}
+			options.StatsBatcher = b
+			batcherDone := make(chan struct{})
+			go func() {
+				b.Run(ctx)
+				close(batcherDone)
+			}()
+			defer func() {
+				<-batcherDone
+			}()
```
Comment (mtojek) [inline cli/server.go:835]: nit: maybe you should consider wrapping into a two-liner: function call with `go func()` inside, and `defer closeFunc`

File: coderd/batchstats/batcher.go [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+	"cdr.dev/slog/sloggers/sloghuman"
+
+	"github.com/coder/coder/coderd/database"
+	"github.com/coder/coder/coderd/database/dbauthz"
+	"github.com/coder/coder/codersdk/agentsdk"
+)
+
+const (
+	// DefaultBatchSize is the default size of the batcher's buffer.
+	DefaultBatchSize = 1024
```
Comment (mtojek) [inline coderd/batchstats/batcher.go:23]: In a typical environment, how many items are in the buffer?

Comment (mtojek) [inline coderd/batchstats/batcher.go:23]: nit: MaxBatchSize?

File: coderd/batchstats/batcher.go [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+	b.buf.RxBytes = append(b.buf.RxBytes, st.RxBytes)
+	b.buf.TxPackets = append(b.buf.TxPackets, st.TxPackets)
+	b.buf.TxBytes = append(b.buf.TxBytes, st.TxBytes)
+	b.buf.SessionCountVSCode = append(b.buf.SessionCountVSCode, st.SessionCountVSCode)
+	b.buf.SessionCountJetBrains = append(b.buf.SessionCountJetBrains, st.SessionCountJetBrains)
+	b.buf.SessionCountReconnectingPTY = append(b.buf.SessionCountReconnectingPTY, st.SessionCountReconnectingPTY)
+	b.buf.SessionCountSSH = append(b.buf.SessionCountSSH, st.SessionCountSSH)
+	b.buf.ConnectionMedianLatencyMS = append(b.buf.ConnectionMedianLatencyMS, st.ConnectionMedianLatencyMS)
+
+	// If the buffer is full, signal the flusher to flush immediately.
```
Comment (mtojek) [inline coderd/batchstats/batcher.go:167]: nit: do you think that we should flush before the buffer is full? Let's say 80% of occupancy?

File: coderd/batchstats/batcher.go [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+	// marshal connections by proto
+	payload, err := json.Marshal(b.connectionsByProto)
+	if err != nil {
+		b.log.Error(ctx, "unable to marshal agent connections by proto, dropping data", slog.Error(err))
+		b.buf.ConnectionsByProto = json.RawMessage(`[]`)
+	} else {
+		b.buf.ConnectionsByProto = payload
+	}
+
+	err = b.store.InsertWorkspaceAgentStats(ctx, b.buf)
```
Comment (mtojek) [inline coderd/batchstats/batcher.go:228]: What happens if one "stat" in the batch is faulty? Is the whole batch rejected?

Bonus points: maybe we need a test for this.

File: coderd/database/queries/workspaceagentstats.sql (hunk 5 above)
Comment (johnstcn) [reply coderd/database/queries/workspaceagentstats.sql:53]: :upside_down_face: 

`jsonb` and not `jsonb[]`

File: coderd/batchstats/batcher.go (hunk 10 above)
Comment (johnstcn) [reply coderd/batchstats/batcher.go:23]: Assuming 1000 workspaces, the worst-case scenario is that they all attempt to push stats at once. That's extremely unlikely however.

File: coderd/coderd.go (hunk 6 above)
Comment (johnstcn) [reply coderd/coderd.go:294]: Good catch, it doesn't need to be.

File: coderd/batchstats/batcher.go (hunk 11 above)
Comment (johnstcn) [reply coderd/batchstats/batcher.go:167]: What would the advantage of this be? If you're trying to avoid blocking during flush, blocking will happen regardless of when we flush. We could make a temporary copy of the buffer while we flush to avoid that, but batch-inserting 1000 entries happens fairly quickly from what I can tell.

File: coderd/batchstats/batcher.go (hunk 12 above)
Comment (johnstcn) [reply coderd/batchstats/batcher.go:228]: Yes, the whole batch will be rejected. That's a good idea for a test.

File: coderd/coderdtest/coderdtest.go (hunk 2 above)
Comment (johnstcn) [reply coderd/coderdtest/coderdtest.go:253]: I expect the latter option, as there would be fewer DB calls overall. I'll go with that one then.

File: coderd/database/dbfake/dbfake.go (hunk 4 above)
Comment (johnstcn) [reply coderd/database/dbfake/dbfake.go:2709]: Yes, the tests added here were the impetus for this change!

**Review by Emyrk — COMMENTED** [review]
Not saying to solve this, but dang our `server.go` start function is growing huge. We should really look into a better way to manage it.

File: coderd/batchstats/batcher.go (hunk 11 above)
Comment (Emyrk) [reply coderd/batchstats/batcher.go:167]: We had a conversation about this.

We came up with passing the existing buffer to the lever and resetting the buffer. So that way the buffer is new for the next `Add` and the previous data is pending the db query.

Psuedo code:
```golang
	if len(b.buf.ID) == cap(b.buf.ID) {
		select{
		case b.flushLever <- struct{}{}:
		default:
			b.log.Error(context.Background(), "this should never happen, dropping agent stats :(")
		}
		b.flushLever <- struct{}{}
		b.flushLever <- &b.buf
		b.buf = make(database.InsertWorkspaceAgentStatsParams, 0, b.batchSize)
	}
```

Some additional nits:
- We should probably use the constant instead of `cap` for comparison since `cap` can grow. `len(b.buf.ID) >= b.batchSize`
- We should use a select and throw an error log if we are backing up. We should drop the stats, as if we are backing up, there is nothing we can do except log it anyway.

File: coderd/batchstats/batcher.go (hunk 12 above)
Comment (mtojek) [reply coderd/batchstats/batcher.go]: It would be nice to exclude the faulty item, and keep the rest of the batch, but I guess that it would be tricky to implement, or just fall back to iterative approach.

**Review by mtojek — APPROVED** [review]
Looks good 👍

File: coderd/batchstats/batcher.go (hunk 11 above)
Comment (johnstcn) [reply coderd/batchstats/batcher.go:159]: It ended up being simpler to do an early flush. However, I thought about this some more and I think using a buffered channel would also resolve this issue. This is a possible later enhancement.

File: cli/server.go (hunk 9 above)
Comment (johnstcn) [reply cli/server.go]: Should be fixed with updated constructor that returns a closer

File: coderd/batchstats/batcher.go [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+	b.buf.TxBytes = append(b.buf.TxBytes, st.TxBytes)
+	b.buf.SessionCountVSCode = append(b.buf.SessionCountVSCode, st.SessionCountVSCode)
+	b.buf.SessionCountJetBrains = append(b.buf.SessionCountJetBrains, st.SessionCountJetBrains)
+	b.buf.SessionCountReconnectingPTY = append(b.buf.SessionCountReconnectingPTY, st.SessionCountReconnectingPTY)
+	b.buf.SessionCountSSH = append(b.buf.SessionCountSSH, st.SessionCountSSH)
+	b.buf.ConnectionMedianLatencyMS = append(b.buf.ConnectionMedianLatencyMS, st.ConnectionMedianLatencyMS)
+
+	// If the buffer is full, signal the flusher to flush immediately.
+	if len(b.buf.ID) == cap(b.buf.ID) {
+		b.flushLever <- struct{}{}
```
[the remark this answers is not in the document]
Comment (johnstcn) [reply coderd/batchstats/batcher.go:167]: This can cause a double-flush, so added an atomic to check if a flush is already in progress.

File: coderd/coderdtest/coderdtest.go (hunk 8 above)
Comment (johnstcn) [reply coderd/coderdtest/coderdtest.go]: Added

File: coderd/database/dbauthz/dbauthz.go (hunk 3 above)
Comment (johnstcn) [reply coderd/database/dbauthz/dbauthz.go:2003]: Fixed

File: cli/server.go (hunk 1 above)
Comment (johnstcn) [reply cli/server.go]: Made the interval 1 second instead by default.

File: coderd/coderd.go (hunk 7 above)
Comment (johnstcn) [reply coderd/coderd.go:296]: No, just keeping to the existing pattern in this file.

File: coderd/batchstats/batcher.go [hunk 14]
```diff
@@ -156,9 +159,13 @@ func (b *Batcher) Add(
...
 
-	// If the buffer is full, signal the flusher to flush immediately.
-	if len(b.buf.ID) == cap(b.buf.ID) {
+	// If the buffer is over 80% full, signal the flusher to flush immediately.
+	// We want to trigger flushes early to reduce the likelihood of
+	// accidentally growing the buffer over batchSize.
+	filled := float64(len(b.buf.ID)) / float64(b.batchSize)
+	if filled >= 0.8 && !b.flushForced.Load() {
 		b.flushLever <- struct{}{}
+		b.flushForced.Store(true)
```
Comment (Emyrk) [inline coderd/batchstats/batcher.go:168]: Does the `flushForced` just prevent a deadlock by adding to the `flushLever` channel if it already has something?

Can we just use a select statement to prevent pushing to the flushLever channel if it already has an item?

Comment (johnstcn) [reply coderd/batchstats/batcher.go:168]: I tried the select but it will still end up doing one extra unwanted flush.

Comment (mtojek) [reply coderd/batchstats/batcher.go:168]: Well, in theory an extra flush is lesser evil than losing data.

**Review by Emyrk — APPROVED** [review]
